ref/haskell/src/Lib.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8)
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Map.HashMap Char Word8
```

Comment:
I could do it for charsetMapInv, but for charsetMap, it would result in a mess. An array needs bounds and be filled continuously within them, but the base32 charset (sorted) is not continuous. This would make the code less elegant, also in base32Decode where I'd need to add explicit charset membership checks.

I could make a newtype and define Ix myself to make it continuous, but that would add a lot of code as well.

Thoughts?